pyomo/contrib/incidence_analysis/config.py

```diff
@@ -54,7 +65,90 @@ class IncidenceMethod(enum.Enum):
 )
 
 
-IncidenceConfig = ConfigDict()
+class _ReconstructVisitor:
+    pass
+
+
+def _amplrepnvisitor_validator(visitor=_ReconstructVisitor):
+    # This checks for and returns a valid AMPLRepnVisitor, but I don't want
+    # to construct this if we're not using IncidenceMethod.ampl_repn.
+    # It is not necessarily the end of the world if we construct this, however,
+    # as the code should still work.
+    if visitor is _ReconstructVisitor:
+        subexpression_cache = {}
+        subexpression_order = []
+        external_functions = {}
+        var_map = {}
+        used_named_expressions = set()
+        symbolic_solver_labels = False
+        # TODO: Explore potential performance benefit of exporting defined variables.
+        # This likely only shows up if we can preserve the subexpression cache across
+        # multiple constraint expressions.
+        export_defined_variables = False
+        sorter = FileDeterminism_to_SortComponents(FileDeterminism.ORDERED)
+        amplvisitor = AMPLRepnVisitor(
+            text_nl_template,
+            subexpression_cache,
+            subexpression_order,
+            external_functions,
+            var_map,
+            used_named_expressions,
+            symbolic_solver_labels,
+            export_defined_variables,
+            sorter,
+        )
+    elif not isinstance(visitor, AMPLRepnVisitor):
+        raise TypeError(
+            "'visitor' config argument should be an instance of AMPLRepnVisitor"
+        )
+    else:
+        amplvisitor = visitor
+    return amplvisitor
+
+
+_ampl_repn_visitor = ConfigValue(
+    default=_ReconstructVisitor,
+    domain=_amplrepnvisitor_validator,
+    description="Visitor used to generate AMPLRepn of each constraint",
+)
```

Comment:
By default, `_ReconstructVisitor` is passed to the validator, and we construct a new visitor. However, as written, cloning this ConfigValue (via `__call__`) will create a new value pointing to the same visitor. This effectively means that, unless a new visitor is explicitly received, the same default visitor is used for all new ConfigValues. I.e. we re-use the visitor for multiple models, or even worse, for the same model before and after modification. The code below implements a hack around `ConfigDict` to re-construct the visitor instead of using the default.